File: src/main/java/unyo/gui/GraphPanel.java
package unyo.gui;

import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;

import javax.swing.JPanel;

import unyo.gui.renderer.VisualGraph;
import unyo.gui.renderer.DefaultRenderer;
import unyo.gui.renderer.GraphicsContext;
import unyo.gui.renderer.DefaultMover;
import unyo.entity.Graph;

@SuppressWarnings("serial")
public class GraphPanel extends JPanel {

    private final VisualGraph visualGraph = new VisualGraph();
    private GraphicsContext graphicsContext = new GraphicsContext();

    private final DefaultMover mover = new DefaultMover();

    public GraphPanel() {
        super(true);

        Thread th = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    while (true) {
                        mover.move(visualGraph);
                        repaint();
                        Thread.sleep(10);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
        th.start();
    }

    public void setGraph(Graph graph) {
        visualGraph.rewrite(graph);
        repaint();
    }

    @Override
    public void paintComponent(Graphics g){
        super.paintComponent(g);

        Graphics2D g2d = (Graphics2D)g;
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        DefaultRenderer r = new DefaultRenderer(g2d, graphicsContext);
        r.render(visualGraph);
    }
}

File: src/main/scala/unyo/gui/GraphPanel.scala
package unyo.gui

import scala.swing.{Panel,Graphics2D}
import scala.swing.event.{MousePressed,MouseReleased,MouseDragged}
import scala.actors.Actor._

import unyo.entity.Graph;
import unyo.util._
import unyo.util.Geometry._
import unyo.Env

class GraphPanel extends Panel() {

  val visualGraph = new VisualGraph
  val graphicsContext = new GraphicsContext
  val mover = new DefaultMover

  listenTo(this.mouse.clicks, this.mouse.moves)
  var prevPoint: java.awt.Point = null
  reactions += {
    case MousePressed(_, p, _, _, _) => prevPoint = p
    case MouseReleased(_, p, _, _, _) => prevPoint = null
    case MouseDragged(_, p, _) => {
      if (prevPoint != null) {
        graphicsContext.moveBy(prevPoint - p)
        prevPoint = p
      }
    }
  }

  actor {
    var prevMsec = System.currentTimeMillis
    loop {
      val msec = System.currentTimeMillis

      mover.move(visualGraph, 1.0 * (msec - prevMsec) / 100)
      repaint

      prevMsec = msec
      Thread.sleep(10)
    }
  }

  def setGraph(g: Graph) {
    visualGraph.rewrite(g)
    repaint
  }

  override def paint(g: Graphics2D) {
    import java.awt.RenderingHints._

    super.paint(g)

    g.setRenderingHint(KEY_ANTIALIASING, VALUE_ANTIALIAS_ON)
    val r = new DefaultRenderer(g, graphicsContext)
    r.render(visualGraph)
  }

}
